UnityServer/functions/src/model/UnityManager.ts

```diff
@@ -0,0 +1,203 @@
+
+    if (purchaseInformation.purchaseType === SkuType.ONE_TIME) {
+      return this.verifier.verifyINAPP(purchaseInformation.receipt).then(() => {
+        return this.verifyPurchaseAndSaveInDB(userId, purchaseInformation.receipt);
+      }).catch((error: any) => {
+        return new UnityStatus(false, error.errorMessage);
+      });
+    } else {
+      return this.verifier.verifySub(purchaseInformation.receipt).then(async () => {
+        const unityStatus = await this.verifyPurchaseAndSaveInDB(userId, purchaseInformation.receipt);
+        if (unityStatus.success) {
+          return this.saveSubscriptionTokenInUsers(userId, purchaseInformation.receipt);
+        }
+        return unityStatus;
+      }).catch((error: any) => {
+        return new UnityStatus(false, error.errorMessage);
+      });
+    }
+  }
+
+  // 2) verification of the purchase token in database and if it's not in the database
+  //    it will save the receipt in purchases table.
+  async verifyPurchaseAndSaveInDB(userId: string, receipt: any): Promise<UnityStatus> {
+    receipt.userId = userId;
+    const tokenCall = this.purchasesCollectionReference.doc(receipt.purchaseToken);
+
+    return tokenCall.get()
+      .then(purchaseToken => {
+        if (!purchaseToken.exists) {
+          return tokenCall.set(receipt)
+            .then(() => {
+              return new UnityStatus(true, "Save in database");
+            }).catch(error => {
+              return new UnityStatus(false, error);
+            });
+        }
+        return new UnityStatus(false, "Purchase token already exists in the database");
+      })
+      .catch(error => {
+        return new UnityStatus(false, error);
+      });
+  }
+
+  //This is an additional save of the subscripton receipt in the users table for
+  // faster access when verifying price change of a subscripton for a user.
+  async saveSubscriptionTokenInUsers(userId: string, receipt: any): Promise<UnityStatus> {
+    return this.usersCollectionReference.doc(userId).collection("subscription").doc("receipt").set(receipt).then(() => {
+      return new UnityStatus(true, "Save in database");
+    }).catch(error => {
+      return new UnityStatus(false, error);
+    });
+  }
+
+  private getReceiptAndPurchaseType(jsonReceipt: string) {
+    const receipt = JSON.parse(this.getReceiptData(jsonReceipt));
+    return { receipt: receipt, purchaseType: this.getPurchaseType(receipt.productId) }
+  }
+
+  // Check if the product id is a subscription or a one time purchase
+  private getPurchaseType(productId: string) {
+    if (productId.includes('subscription')) {
+      return SkuType.SUBS;
+    }
+    else {
+      return SkuType.ONE_TIME;
+    }
+  }
+
+  // Editing the receipt for parsing it in json
+  private getReceiptData(receipt: string) {
+    let receiptInfo = receipt;
+    receiptInfo = receiptInfo.replace(/\\/g, '');
+    return receiptInfo.slice(receiptInfo.search("orderId") - 2, receiptInfo.search("signature") - 3);
+  }
+}
```

Comment:
this receipt itself is a JSON data, you can parse it to get receiptData.

Then receiptData is another JSON and you can parse it again.

Reply:
Yep, was overthinking it 😅.Going to fix it